src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -27,3 +36,792 @@ def deduce_line_indent(raw_segment: RawSegment, root_segment: BaseSegment) -> st
+        last_line_idx = group_lines[-1]
+        intermediate_lines = [
+            line
+            for line in lines[first_line_idx + 1 : last_line_idx]
+            # Exclude lines which are in the group to avoid
+            # issues with loop markers.
+            if line not in [lines[idx] for idx in group_lines]
+        ]
+        reflow_logger.debug(
+            "    Intermediate Lines: %s",
+            [line.initial_indent_balance for line in intermediate_lines],
+        )
+        limit_indent = min(
+            # Minus one to reverse the effect that the block has
+            # already had.
+            line.initial_indent_balance - 1
+            for line in intermediate_lines
+        )
+
+        # Evaluate options.
+        overlap = set.intersection(*options)
+        reflow_logger.debug("    Simple Overlap: %s", overlap)
+        # Remove any options above the limit option.
+        # We minus one from the limit, because if it comes into effect
+        # we'll effectively remove the effects of the indents between the elements.
+        overlap = {i for i in overlap if i <= limit_indent}
+        reflow_logger.debug("    Overlap: %s, Limit: %s", overlap, limit_indent)
+        # Is there a mutually agreeable option?
+        if overlap:
+            # Go for the deeper option if there's flexibility, because this
+            # will usually involve moving the fewest options.
+            best_indent = max(overlap)
+            reflow_logger.debug(
+                "    Case 2: Best: %s, Overlap: %s", best_indent, overlap
+            )
+        # If no overlap, it's case 3
+        else:
+            # Set the indent to the minimum of the existing ones.
+            best_indent = min(lines[idx].initial_indent_balance for idx in group_lines)
+            reflow_logger.debug("    Case 3: Best: %s", best_indent)
+            # Remove one indent from all intermediate lines.
+            # This is because we're effectively saying that these
+            # placeholders shouldn't impact the indentation within them.
+            for idx in range(first_line_idx + 1, last_line_idx):
+                if idx not in group_lines:
+                    # MUTATION
+                    lines[idx].initial_indent_balance -= 1
+
+        # Set all the lines to this indent
+        for idx in group_lines:
+            # MUTATION
+            lines[idx].initial_indent_balance = best_indent
+
+
+def _revise_comment_lines(lines: List[_IndentLine], elements: ReflowSequenceType):
+    """Given an initial set of individual lines. Revise comment ones.
+
+    NOTE: This mutates the `lines` argument.
+
+    We do this to ensure that lines with comments are aligned to
+    the following non-comment element.
+    """
+    reflow_logger.debug("# Revise comment lines.")
+    # new_lines: List[_ReindentLine] = []
+    comment_line_buffer: List[int] = []
+
+    # Slice to avoid copying
+    for idx, line in enumerate(lines[:]):
```

Comment:
How is this different from just `enumerate(lines)`? As written, it constructs a shallow copy of `lines`, but in either case, the underlying list elements are the same.

Reply:
I'm doing a shallow copy, because I'm editing `lines` within the loop. It felt like bad practice to modify any object which I'm directly iterating over without doing a shallow copy. I'll add a comment.

Reply:
I saw that other loops were editing `lines` in the loop. At a glance, I didn't notice that here. Definitely making a copy is good if you're adding or deleting items. Probably even if just replacing items, although that may technically work without it. 👍